ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm

```diff
@@ -263,4 +275,114 @@ - (void)testShouldHandleVnodeOpEvent {
+    XCTAssertTrue(
+        ShouldHandleFileOpEvent(
+            &perfTracer,
+            context,
+            testVnodeFile.get(),
+            KAUTH_FILEOP_OPEN,
+            false, // isDirectory,
+            &rootHandle,
+            &vnodeFsidInode,
+            &pid));
+    XCTAssertTrue(rootHandle == testRootHandle);
+    // Finding the root should have added testVnodeFile to the cache
+    XCTAssertTrue(testVnodeFile.get() == cacheWrapper[ComputeVnodeHashIndex(testVnodeFile.get())].vnode);
+    XCTAssertTrue(rootHandle == cacheWrapper[ComputeVnodeHashIndex(testVnodeFile.get())].virtualizationRoot);
+    
+    // KAUTH_FILEOP_LINK
+    XCTAssertTrue(
+        ShouldHandleFileOpEvent(
+            &perfTracer,
+            context,
+            testVnodeFile.get(),
+            KAUTH_FILEOP_LINK,
+            false, // isDirectory,
+            &rootHandle,
+            &vnodeFsidInode,
+            &pid));
+    XCTAssertTrue(rootHandle == testRootHandle);
+    // KAUTH_FILEOP_LINK should invalidate the cache entry for testVnodeFile
+    XCTAssertTrue(testVnodeFile.get() == cacheWrapper[ComputeVnodeHashIndex(testVnodeFile.get())].vnode);
+    XCTAssertTrue(RootHandle_Indeterminate == cacheWrapper[ComputeVnodeHashIndex(testVnodeFile.get())].virtualizationRoot);
+    
+    // KAUTH_FILEOP_RENAME (file)
+    // Set a different value in the cache for testVnodeFile's root to validate that the cache is refreshed for renames
+    cacheWrapper[ComputeVnodeHashIndex(testVnodeFile.get())].virtualizationRoot = rootHandle + 1;
+    XCTAssertTrue(
+        ShouldHandleFileOpEvent(
+            &perfTracer,
+            context,
+            testVnodeFile.get(),
+            KAUTH_FILEOP_RENAME,
+            false, // isDirectory,
+            &rootHandle,
+            &vnodeFsidInode,
+            &pid));
+    XCTAssertTrue(rootHandle == testRootHandle);
+    // The cache should have been refreshed for KAUTH_FILEOP_RENAME
+    XCTAssertTrue(testVnodeFile.get() == cacheWrapper[ComputeVnodeHashIndex(testVnodeFile.get())].vnode);
+    XCTAssertTrue(rootHandle == cacheWrapper[ComputeVnodeHashIndex(testVnodeFile.get())].virtualizationRoot);
+    
+    // KAUTH_FILEOP_RENAME (directory)
+    // Directory KAUTH_FILEOP_RENAME events should invalidate the entire cache and then insert
+    // only the directory vnode into the cache
+    cacheWrapper.FillAllEntries();
+    XCTAssertTrue(
+        ShouldHandleFileOpEvent(
+            &perfTracer,
+            context,
+            testVnodeDirectory.get(),
+            KAUTH_FILEOP_RENAME,
+            true, // isDirectory,
+            &rootHandle,
+            &vnodeFsidInode,
+            &pid));
+    XCTAssertTrue(rootHandle == testRootHandle);
+    
+    // Validate the cache is empty except for the testVnodeDirectory entry
+    uintptr_t directoryVnodeHash = ComputeVnodeHashIndex(testVnodeDirectory.get());
+    for (uintptr_t index = 0; index < cacheWrapper.GetCapacity(); ++index)
```

Comment:
To me it looks like this checks the index that should be a hit, and then continues until capacity making sure the rest is empty.  Should it loop around and start at 0 until it hits the index?

Reply:
> Should it loop around and start at 0 until it hits the index?

@jeschu1, I might be missing something in your comment.  The loop starts at index 0 and checks every index in the cache.  If the index equals `directoryVnodeHash` we validate that the entry has the contents we expect, and otherwise we validate that the entry is missing.

Did you see that we were sometimes missing the start of the cache when validating?